presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -442,25 +447,51 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Str
         // last page is removed.  If another thread observes this state before the response is cached
         // the pages will be lost.
         Iterable<List<Object>> data = null;
+        List<String> binaryData = null;
         try {
-            ImmutableList.Builder<RowIterable> pages = ImmutableList.builder();
-            long bytes = 0;
             long rows = 0;
+            long bytes = 0;
             long targetResultBytes = targetResultSize.toBytes();
-            while (bytes < targetResultBytes) {
-                SerializedPage serializedPage = exchangeClient.pollPage();
-                if (serializedPage == null) {
-                    break;
+            if (binaryResults) {
+                ImmutableList.Builder<String> pages = ImmutableList.builder();
```

Comment:
Nit: How do you feel about merging these two of the loops ? 

Also is it expensive to recreate DynamicSliceOutput(1000) each time in the loop ? No idea, so just asking :-)